### ISSUE TITLE
Most users will experience a no-matching-credential dead-end

### ISSUE THREAD

**Issue by dickhardt** [w3c-fedid/digital-credentials#104]
Many users, and all users initially, won't have any wallets installed on their device.  

Letting an app know if there are any wallets lets the app know if the user understands what a wallet is and to not use the API at all. 

This single bit of data discloses if functionality will exist.

**Comment by RByers** [comment]
Agreed we should explore APIs with a single bit of privacy impact. Another option is "has this user ever presented a digital credential before". This is probably sensitive right now given how few users would have either of these bits. But perhaps once we've seen a bit of non-trivial deployment we could explore this. I agree it'll be important for widespread adoption to give verifiers some way to minimize the no-credential friction while also having prominent UI in the likely success case.

**Comment by dickhardt** [comment]
Rather than having an API that can be used for gathering 1 bit of data, it would have the same result if invoking the API was a UX no-op if the user had never installed a credential. The app is then learning after the fact if UX was presented -- and UX would be presented even if the request could not be satisfied, but the user had added at least one credential. 

This way we don't confuse users that don't understand what a credential is.

**Comment by RByers** [comment]
From meeting discussion this evening: updating the issue title to reflect the problem rather than just one possible solution. 

We've heard from multiple potential verifiers that it's important to them that the "no matching credentials" case tends to be low-friction, especially early when very few users will have matching credentials. They won't adopt an API which adds friction but no value in the majority of cases. But this is in conflict with the privacy requirement not to divulge any information to websites without user permission.

> Rather than having an API that can be used for gathering 1 bit of data, it would have the same result if invoking the API was a UX no-op if the user had never installed a credential. The app is then learning after the fact if UX was presented -- and UX would be presented even if the request could not be satisfied, but the user had added at least one credential.

Yeah we've explored that (it was actually our original chromium implementation). The problem is that it's trivial for a site to make a pretty reliable prediction on whether UX was shown by watching the input events that occur on the page. An attacker could invoke a series of requests and watch to see when input behavior changed to indicate the bottom of the screen was now covered in order to learn which sort of credential the user has without the user's permission.

@npdoty makes the good suggestion that perhaps we can address this problem by having the user initiative the presentation rather than the site. For example maybe browsers could have show a button in the URL bar when the page supports digital credentials and the user can choose to tap it. This, of course, has the downside of likely going unnoticed in many cases. But there's a variety of UX design options here.

**Comment by dickhardt** [comment]
> Yeah we've explored that (it was actually our original chromium implementation). The problem is that it's trivial for a site to make a pretty reliable prediction on whether UX was shown by watching the input events that occur on the page. An attacker could invoke a series of requests and watch to see when input behavior changed to indicate the bottom of the screen was now covered in order to learn which sort of credential the user has without the user's permission.

If UX was shown then the site did not gather data silently. I'm assuming the user has to dismiss the UX.  

My suggestion was that the UX is always shown if the user has any credential, so this does not let an attacker learn which credential, it just lets them know there is a credential, and the user also has learned the site has requested a credential. It is only silent if the user has no credential.

**Comment by npdoty** [comment]
As credentials can be obtained and used in so many ways, does the single bit of information that a user has a credential of some kind, or a wallet app of some kind, help very much with the dead-end problem?

This doesn't seem like a UI that should be triggered frequently, rapidly or without significant set-up around the action.

**Comment by dickhardt** [comment]
It certainly helps in the near future when no one has a credential. I would be very reluctant to use the API if 99.99% of the time no one has a credential. And why will someone get a credential if 99.99% of all the sites don't support getting one?

I mind less someone getting UI when they don't have a credential, but have added one. But I don't want to confuse the user that has never added one and does not know what they are. 

Depending on what credentials actually get traction, and user's understanding of them, I want more bits of data before I invoke an API that shows UI, but in the near term, it is super unlikely I will invoke the API.

**Comment by hlflanagan** [comment]
Suggest we look at https://datatracker.ietf.org/doc/html/draft-thomson-postel-was-wrong-03 to see how it might apply.

**Comment by npdoty** [comment]
For government-issued credentials with high-assurance personal information, this shouldn't generally be a quick or automatic prompt, or something done on many webpages. Instead, the goal should be that a user has been given a lot of context, and decided to take an intentional action to present an identity document or assertion, and wanting to do so with a digital credential wallet. Given that level of context, I'm not sure I see the size of the marginal benefit to having a button hidden or not.

**Comment by dickhardt** [comment]
I'm not following what your point is @npdoty - would you elaborate?

**Comment by npdoty** [comment]
(Apologies, forgot the follow-up request until now.) My thinking that because this request was likely to be high-context, explained with some alternatives and not something that would be prompted automatically without a considered user decision, that signals about availability of a wallet or a credential would be less necessary for automatically configuring UI, like showing or hiding a button.

**Comment by Sakurann** [comment]
what about issuance during presentation scenario? where the wallet is capable of handling credential A, the user does not have that credential in the wallet, but has a wallet installed and is willing to get credential issued into it and than presenting it in response to the presentation request. I assume it could help with adoption if this scenario is supported.

**Comment by marcoscaceres** [comment]
I think this is outside the scope of the specification. As @Sakurann pointed out, if no credential is available, the platform or a  wallet could theoretically intervene offer to issue a credential, which, if provisioned in a timely manner, could then be returned as part of the request. 

If you also consider how some platforms handle this for payment, some OSs allow the user to enroll a payment instrument during the  OS setup process. Government-issued credentials are not really dissimilar to payment instruments in that respect. 

Point being: implementers should have the ability to innovate in this space, without the potential privacy leak of revealing anything about what wallets are installed or if there are any credentials available in those wallets.

**Comment by marcoscaceres** [comment]
@dickhardt, is it ok to close this for now and let user agents experiment dealing with this?

**Comment by marcoscaceres** [comment]
This issue has been marked `pending closure`. The spec deliberately accepts the no-matching-credential dead-end as a privacy tradeoff, and `userAgentAllowsProtocol()` was added for protocol-level feature detection.

@dickhardt, any objections to closing this? If we don't hear back within two weeks (by May 14), we'll close it.

**Comment by hlflanagan** [comment]
Brief discussion during the [4 May 2026 meeting](https://github.com/w3c-fedid/meetings/blob/main/2026/2026-05-04-DCAPI-A-notes.md#aob)

**Comment by hlflanagan** [comment]
I haven't seen further updates, so closing. We can reopen if new information becomes available.
